src/python/pants/pantsd/pants_daemon.py

```diff
@@ -125,11 +125,18 @@ def _setup_services(build_root, bootstrap_options, legacy_graph_helper, watchman
 
       :returns: A tuple of (`tuple` service_instances, `dict` port_map).
       """
+      # A lock to ensure that no interactions that require native Scheduler threads or acquire
+      # native locks happen during a fork. Acquired:
+      #   1) During graph warming or filesystem invalidation (by SchedulerService)
+      #   2) During forking (by DaemonPantsRunner via PailgunService)
+      fork_lock = threading.RLock()
+
       fs_event_service = FSEventService(watchman, build_root, bootstrap_options.pantsd_fs_event_workers)
-      scheduler_service = SchedulerService(fs_event_service, legacy_graph_helper)
+      scheduler_service = SchedulerService(fork_lock, fs_event_service, legacy_graph_helper)
```

Comment:
should use the existing `PantsService.lock` here vs plumbing/introducing a new one x2

Reply:
I removed it.